posts/2018-08-07-refactoring-monads.md

```diff
@@ -0,0 +1,391 @@
+function does a little work or some error checking, and then calls the
+next appropriate function in the chain.
+
+You can imagine what kind of code will result. The functions are
+smaller, but it's hard to describe what each function does, and
+functions occasionally have to carry along additional parameters that
+they will ignore except to pass deeper into the call chain. Let's take
+a look at an example refactoring:
+
+```tut:silent
+import scala.util.{Failure, Success, Try}
+
+def extractEBook(
+    id: Int,
+    title: String,
+    author: String,
+    downloadTypeStrOpt: Option[String]): Try[EBook] =
+  downloadTypeStrOpt match {
+    case None => Failure(new AssertionError())
+    case Some(downloadTypeStr) =>
+      DownloadType.fromString(downloadTypeStr) match {
+        case None =>
+          Failure(new ParseError(downloadTypeStr))
+        case Some(dt) =>
+          Success(EBook(id, title, author, dt))
+      }
+  }
+
+def extractBook(
+    id: Int,
+    title: String,
+    author: String,
+    formatStr: String,
+    downloadTypeStrOpt: Option[String]): Try[Book] =
+  Format.fromString(formatStr) match {
+    case None => Failure(new ParseError(formatStr))
+    case Some(Print) =>
+      Success(PrintBook(id, title, author))
+    case Some(Digital) =>
+      extractEBook(id, title, author, downloadTypeStrOpt)
+  }
+
+def findBookById(id: Int): Try[Book] =
+  DB.unsafeQueryUnique(sql"""select * from catalog where id = $id""").flatMap { row =>
+    val id = row[Int]("id")
+    val title = row[String]("title")
+    val author = row[String]("author")
+    val formatStr = row[String]("format")
+    val downloadTypeStr = row[Option[String]]("download_type")
+    extractBook(id, title, author, formatStr, downloadTypeStr)
+  }
+```
+
+As you can see, this form has more manageably-sized functions,
+although they are still a little long. You can also see that the flow
+of control is distributed through all three functions, which means
+understanding the logic enough to modify or test it requires
+understanding all three functions both individually and as a whole. To
+follow the logic, we must trace the functions like a recursive descent
+parser.
+
+### Refactoring with Monads
+Without throwing exceptions and catching them at the top, it's going
+to be hard to do substantially better than the "tail-refactoring"
+approach, unless we start to make use of the fact that we're working
+with `Try`, a data type that supports `flatMap`. More precisely, `Try`
+has a monad - recall that monads let us model computational effects
+that take place in sequence.
+
+Let's try to factor out smaller functions, each returning `Try`, and
+then use a for-comprehension to specify the sequence of operations:
+
+```tut:silent
+import scala.util.{Failure, Success, Try}
+
+def tryParse[A](s: String, parse: String => Option[A]): Try[A] = {
+  parse(s)
+    .map(Success(_))
+    .getOrElse(Failure(new ParseError(s)))
+}
```

Comment:
Wouldn't this be easier by having the parse functions return `Try` directly instead of `Option`? I was actually already wondering this above, where `ParseError` is defined but not immediately used.

Reply:
Yes, I think so.

We tend to use a return type of `Option` when writing something like `.fromString` because it's a little lighter weight, isn't opinionated as to whether a missing value is an error or not, and it's easy to convert to `Try` with an appropriate exception type later with adapters like `tryParse`. I did the same here out of habit.

I'll switch the `fromString` methods over to something that returns `Try` instead, so we can get rid of `tryParse` entirely. The example should still hold up and there'll be less code for people to grok.

Reply:
Perfect, thanks!